### PR TITLE
BENCH: Add benchmarks for signal.find_peaks and related functions

### LINKED ISSUE #8601: BENCH: Benchmarks for find_peaks, peak_prominences & peak_widths
As suggested in #8594 it might be worthwhile to add asv benchmarks for the recently added functions

- [ ] [find_peaks](https://scipy.github.io/devdocs/generated/scipy.signal.find_peaks.html)
- [ ] [peak_prominences](https://scipy.github.io/devdocs/generated/scipy.signal.peak_prominences.html)
- [ ] [peak_widths](https://scipy.github.io/devdocs/generated/scipy.signal.peak_widths.html)

introduced with #8264. This may act as a reminder and place of discussion until I or someone else takes this up in a PR or it is resolved by other means.

See also: #8541, #8523

### PR DESCRIPTION
**Pull request by lagru**
These benchmarks use `scipy.signal.electrocardiogram` as input. I figured that a real (bio)signal provides a more realistic use case in comparison to randomly generated synthetic signals, e.g. `np.random.rand`.

The currently provided benchmarks try to isolate the algorithms thus each benchmark only evaluates one of the following tasks: maxima detection, distance filtering, prominence calculation & width calculation.

If you think more scenarios would be useful please let me know.

Closes #8601

### CODE REVIEW

File: benchmarks/benchmarks/peak_finding.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+from .common import Benchmark
+
+
+class FindPeaks(Benchmark):
+    """Benchmark `scipy.signal.find_peaks`.
+
+    Notes
+    -----
+    The first value of `distance` is None in which case the benchmarks shows
```
Comment (tylerjereddy) [inline benchmarks/benchmarks/peak_finding.py]: Plural check on "benchmarks shows" --> "benchmark shows" maybe?

**Review by tylerjereddy — COMMENTED** [review]
These `asv` benchmarks look mechanically sound; perhaps good for a `signal` guru to have a quick look as well.

Did you try to do a quick asv run locally for issues? I know it can be a bit of pain to get the benchmarks running sometimes.

**Comment by lagru** [comment]
> Did you try to do a quick asv run locally for issues? I know it can be a bit of pain to get the benchmarks running sometimes.

Yeah, I did but only with `python runtests.py --bench "peak_finding.*"`. I started a job with `asv` itself too but didn't wait for it to complete. Building the virtualenvs and running the benchmarks takes really long on my notebook. Is 1h per commit normal?

**Comment by pv** [comment]
@lagru: maybe you did `asv run` and set it up to run benchmarks for all commits in scipy?

**Comment by lagru** [comment]
The exact command I used was the recommended one `python run.py run --skip-existing-commits --steps 10 ALL` so it should have build maximal 10 commits if I understand correctly. Anyway even doing one commit seemed slow. I think that's just because my notebook's CPU isn't that powerful...

**Comment by pv** [comment]
You have to measure for about ~1 sec per benchmark to get useful timings (if you have hardware set up correctly --- on laptops this still gives bad results), and there are ~2000 benchmarks (many parameter combinations), so it boils down to about half an hour per commit.

**Comment by lagru** [comment]
Ah, thanks for the explanation. That does sound reasonable. :)

**Review by tylerjereddy — APPROVED** [review]
Ok, +1 to merge now on my end.

**Comment by larsoner** [comment]
@tylerjereddy I know the `signal` end but not the benchmark/asv end. How do I evaluate? Do we have docs about this somewhere?

**Comment by tylerjereddy** [comment]
@larsoner 

We don't have `asv` benchmark running code built directly into our CI matrix, so that's why I was asking if they had been checked locally. I think it just takes far too long to run to build into CI.

From a signal expertise standpoint, maybe just check the sizes / values of the parameters used to see if anything looks fishy & check that the functions being tested are the pertinent ones. 

The layout of the benchmarks is similar to a unit test in many ways, with parameters in setup functions initialized and then the timing of the focused benchmark is isolated in a small timing function which accepts the various parameters (it will use all possible combinations of parameters provided).

**Comment by larsoner** [comment]
> I think it just takes far too long to run to build into CI.

IIRC CircleCI is built for long jobs. If you mean 30-60 minutes this seems tolerable. We could always run it on:

1. All commits to `master`
2. On PRs that touch certain files (maybe any under `benchmarks`)
3. Any PR commit that contains some custom magic message like `[circleci benchmark]`.

These are not too difficult to do in the CircleCI file, we do this sort of thing in [MNE-Python](https://github.com/mne-tools/mne-python/blob/master/.circleci/config.yml#L56). If you think it's worthwhile, I am happy to help set it up.

**Comment by larsoner** [comment]
Thanks @lagru !

**Comment by lagru** [comment]
Thank you for reviewing.

**Pull request merged**